lib/collections.c

```diff
@@ -344,77 +377,61 @@ void dictionary_set(Dictionary* dictionary, const char* key, void* value)
             exit(EXIT_FAILURE);
         }
         KeyValuePair key_value_pair = {copied_key, value};
-        dictionary_rebuild(dictionary);
-        index = hash_key(key) % dictionary->max_length;
-        while (dictionary->key_value_pairs[index].key != NULL)
-        {
-            index = (index + 1) % dictionary->max_length;
-        }
+        dictionary_rebuild(&dictionary);
+        index = dictionary_lookup(dictionary, key);
         dictionary->key_value_pairs[index] = key_value_pair;
         dictionary->length++;
     }
+    // Key already exists, just replace value (no need to free the key)
     else
     {
-        while (dictionary->key_value_pairs[index].key != NULL)
+        if (strcmp(dictionary->key_value_pairs[index].key, key) == 0)
```

Comment:
I think this condition is *implied* by the spec of `dictionary_lookup`, so it should be unnecessary.

optional: you could make it an assert, to signal intent to the reader